DOC Update the FAQ entry on GPU support (#28328)
scikit-learn/scikit-learn: doc/faq.rst

Will you add GPU support?
^^^^^^^^^^^^^^^^^^^^^^^^^

A limited but growing list of scikit-learn estimators can already run on GPUs

> via the :ref:`array_api` on PyTorch of CuPy input data for instance. Most

Author: @betatim thanks. I tried something along those lines in 43c6fe0.

> via the :ref:`array_api` on PyTorch of CuPy input data for instance. Most

Reviewer: LGTM

estimators that rely on NumPy for their computational intensive operations can
be considered for Array API support and therefore GPU support.

However, not all scikit-learn estimators are amenable to running on GPUs via
the Array API for fundamental algorithmic reasons. For instance, tree-based
models currently implemented with Cython in scikit-learn are fundamentally not

> array-based algorithms. Other algorithms such as k means or k nearest neighbors

Reviewer 2: Yeah, fair-ish 😁

> array-based algorithms. Other algorithms such as k means or k nearest neighbors

Reviewer 3: I would keep the original
> The main reason is that GPU support will introduce many software dependencies and introduce platform specific issues.

refer to specialized NN and tree-based libs and then continue with the text on Cython.

> array-based algorithms. Other algorithms such as k means or k nearest neighbors

Author: I am not sure to see how that would work. Can you please comment with the full set of sentences you have in mind?

> array-based algorithms. Other algorithms such as k means or k nearest neighbors

Author: Note: the end of the paragraph already mentions not adding GPU specific dependencies to the base scikit-learn package.

rely on array-based algorithms but cannot be efficiently implemented without
fusing consecutive operations to avoid introducing performance killing memory
access to unnecessarily large intermediate arrays. Fusing such operations
cannot be expressed via the Array API for the foreseeable future. Adding
efficient GPU support to these estimators would require designing and adopting

> a more flexible extension system for scikit-learn. This possibility is being
> considered in the following GitHub issue (under discussion):
> 
> - https://github.com/scikit-learn/scikit-learn/issues/22438

Author: I am ok to remove the link to the issue to manage expectations w.r.t. this particular point.

```suggestion
a more flexible extension system for scikit-learn.
```

Reviewer: I think we should include the link if we want people to join the discussion. Which is probably fine, and we can always remove it if we get too many low quality contributions.

Overall +0 from me for keeping it.

Will you add GPU support?
^^^^^^^^^^^^^^^^^^^^^^^^^

> A limited but growing :ref:`list of scikit-learn estimators

Reviewer 3: ```suggestion
Since 20XX, a limited but growing :ref:`list of scikit-learn estimators
```
Having a year would make clear that this is a recent change.

<array_api_supported>` can already run on GPUs if the input data is provided
as a PyTorch or CuPy array and if scikit-learn has been configured to accept
such inputs as explained in :ref:`array_api`.

> Most estimators that rely on NumPy for their computational intensive operations

Author: ```suggestion
Most estimators that rely on NumPy for their computationally intensive operations
```

can be considered for Array API support and therefore GPU support.

> However, not all scikit-learn estimators are amenable to running on GPUs via

Reviewer: I'm wondering if we need the level of detail of this paragraph. Maybe we can just state that "However, not all scikit-learn estimators can be efficiently implemented using the Array API. Adding efficient GPU support to these estimators would require designing and adopting a more flexible extension system for scikit-learn. This possibility is currently being considered."

I'm thinking that we might not need to explain why this is so in a FAQ entry. I assume most people looking at the FAQ are here for a quick and simple answer. The longer we make the answer the more things there are that people could misunderstand or maybe even disagree with (and then start arguing over whether or not they are true).

What do you think?

Author: As you prefer. The me from 3 years ago (before I had the opportunity to learn more about those matters) would have appreciated a finer level of details to understand why the short version you propose is not a completely arbitrary statement.

Reviewer: Fair point. Not sure I have a strong opinion, I like shorter text but I can see that it is interesting to know why the short version is what it is.

Author: @adrinjalali do you want to be the tie breaker on this?

Reviewer 2: It was me, I would go for even a longer version. So I'd say this is a good balance.

the Array API for fundamental algorithmic reasons. For instance, tree-based
models currently implemented with Cython in scikit-learn are fundamentally not
array-based algorithms. Other algorithms such as k means or k nearest neighbors
rely on array-based algorithms but cannot be efficiently implemented without
fusing consecutive operations to avoid introducing performance killing memory
access to unnecessarily large intermediate arrays. Fusing such operations
cannot be expressed via the Array API for the foreseeable future. Adding
efficient GPU support to these estimators would require designing and adopting
a more flexible extension system for scikit-learn. This possibility is being
considered in the following GitHub issue (under discussion):

- https://github.com/scikit-learn/scikit-learn/issues/22438

> In any case, the `scikit-learn` package itself will not add any mandatory

Reviewer: ```suggestion
The `scikit-learn` package itself will not add any mandatory
```

IMHO "In any case" doesn't add anything/just adds more words. We could go even further and turn the sentence around to "No mandatory GPU-specific software dependencies will be added to the `scikit-learn` package to ensure that it remains lightweight and easy to install." - this puts the important message at the start of the sentence.

Will you add GPU support?
^^^^^^^^^^^^^^^^^^^^^^^^^

Adding GPU support by default would introduce heavy harware-specific software

> dependencies which make it both harder for the average user to install

Reviewer 4: and reimplementation of algorithms? No dependency can make our trees run on GPU as explained below, right?

Author: Indeed, the sentence was updated based on Christian's suggestion to address this.

> dependencies which make it both harder for the average user to install
> scikit-learn and harder for the developers to maintain the code.

Reviewer 3: ```suggestion
dependencies and existing algorithms would need to be reimplemented. This would make it both harder for the average user to install
scikit-learn and harder for the developers to maintain the code.
```

However, since 2023, a limited but growing :ref:`list of scikit-learn
estimators <array_api_supported>` can already run on GPUs if the input data is
provided as a PyTorch or CuPy array and if scikit-learn has been configured to
accept such inputs as explained in :ref:`array_api`. This Array API support
allows scikit-learn to run on GPUs without introducing heavy and
hardware-specific software dependencies to the main package.

Most estimators that rely on NumPy for their computationally intensive operations
can be considered for Array API support and therefore GPU support.

However, not all scikit-learn estimators are amenable to efficiently running
on GPUs via the Array API for fundamental algorithmic reasons. For instance,
tree-based models currently implemented with Cython in scikit-learn are

> fundamentally not array-based algorithms. Other algorithms such as k means or k
> nearest neighbors rely on array-based algorithms but are also implemented in

Reviewer 3: ```suggestion
fundamentally not array-based algorithms. Other algorithms such as k-means or
k-nearest neighbors rely on array-based algorithms but are also implemented in
```

Cython in order to manually interleave consecutive array operations to avoid

> introducing performance killing memory access to large intermediate arrays:

Reviewer 3: Can this sentence be simplified or broken down?

Author: Done.
